### PR TITLE
edgeql: Disallow nested arrays.

### PR DESCRIPTION
**Pull request by vpetrovykh**
Currently nested arrays are disallowed in EdgeDB, so these tests should
all come up with errors.

### CODE REVIEW

**Comment by elprans** [comment]
The fix for this would be to prohibit passing `array` to any function declared as `func(any) -> array<any>`. This can be done in the function handler of the IR compiler.

File: edb/lang/edgeql/compiler/func.py [hunk 1]
```diff
@@ -63,8 +63,10 @@ def compile_FunctionCall(
         fctx.in_func_call = True
         args, kwargs, arg_types = process_func_args(expr, funcname, ctx=fctx)
 
+        fatal_array_check = len(funcs) == 1
```
Comment (1st1) [inline edb/lang/edgeql/compiler/func.py]: I don't like this name. How about `nested_array_check`

File: edb/lang/edgeql/compiler/func.py [hunk 2]
```diff
@@ -133,6 +139,24 @@ def check_function(
...
+            # 'array<any>', and the argument is also an 'array', then
+            # this is an invalid function invocation.
+            rt = func.returntype
+
+            if (pt.name == 'std::any' and
+                    isinstance(rt, s_types.Array) and
+                    rt.get_subtypes()[0].name == 'std::any' and
+                    isinstance(at, s_types.Array)):
+
+                if fatal_array_check:
```
Comment (1st1) [inline edb/lang/edgeql/compiler/func.py]: Re-arrange IFs:

```python
if (pt.name == 'std::any' and
        fatal_array_check and
        ...)
```

File: tests/test_edgeql_functions.py [hunk 3]
```diff
@@ -279,6 +281,17 @@ class TestEdgeQLFunctions(tb.QueryTestCase):
             [[{'name': 'Elvis'}, {'name': 'Yury'}]]
         ])
 
+    async def test_edgeql_functions_array_agg_13(self):
+        # FIXME: please update the error message once an error is
+        # actually produced
```
Comment (1st1) [inline tests/test_edgeql_functions.py]: Outdated comment?

File: tests/test_edgeql_expressions.py [hunk 4]
```diff
@@ -796,6 +796,44 @@ class TestExpressions(tb.QueryTestCase):
             [],
         ])
 
+    @unittest.expectedFailure
+    async def test_edgeql_expr_array_10(self):
+        # FIXME: please update the error message once an error is
+        # actually produced
+        with self.assertRaisesRegex(exc.EdgeQLError, 'nested array'):
```
Comment (elprans) [inline tests/test_edgeql_expressions.py]: This should be easy to fix in the array literal handler.

File: tests/test_edgeql_expressions.py [hunk 5]
```diff
@@ -796,6 +796,44 @@ class TestExpressions(tb.QueryTestCase):
...
+                SELECT [[1, 2], [3, 4]];
+            ''')
+
+    @unittest.expectedFailure
+    async def test_edgeql_expr_array_11(self):
+        # FIXME: please update the error message once an error is
+        # actually produced
+        with self.assertRaisesRegex(exc.EdgeQLError, 'nested array'):
+            await self.con.execute(r'''
+                SELECT [array_agg({1, 2})];
```
Comment (elprans) [inline tests/test_edgeql_expressions.py]: Ditto

**Comment by vpetrovykh** [comment]
OK, I've updated the literal handler to check for nested arrays.

**Pull request merged**